egs/wsj/s5/steps/nnet3/multilingual/allocate_multilingual_examples.py

```diff
@@ -0,0 +1,264 @@
+
+  # If weights are not provided, the scaling weights are 1.0.
+  if args.lang2weight is None:
+    lang2weight = [ 1.0 ] * num_langs
+  else:
+    lang2weight = ReadLang2Len(args.lang2weight)
+    assert(len(lang2weight) == num_langs)
+
+  if not os.path.exists(args.egs_dir + "/temp"):
+    os.makedirs(args.egs_dir + "/temp")
+
+  num_lang_file = open(args.egs_dir + "/info/" + args.prefix + "num_tasks", "w");
+  print("{0}".format(num_langs), file = num_lang_file)
+
+
+  # Each element of all_egs (one per num_archive * num_jobs) is
+  # an array of 3-tuples (lang-id, local-start-egs-line, num-egs)
+  all_egs = []
+  lang_len = lang2len[:]
+  tot_num_egs = sum(lang2len[i] for i in range(len(lang2len))) # total num of egs in all languages
+  num_archives = max(1, min(args.max_archives, tot_num_egs / args.samples_per_iter))
+
+
+  num_arch_file = open(args.egs_dir + "/info/" + args.prefix + "num_archives", "w");
+  print("{0}".format(num_archives), file = num_arch_file)
+  num_arch_file.close()
+
+  this_num_egs_per_archive = tot_num_egs / (num_archives * args.num_jobs) # num of egs per archive
+  for job_index in range(args.num_jobs):
+    for archive_index in range(num_archives):
+      # Temporary scp.job_index.archive_index files to store egs.scp correspond to each archive.
+      print("Processing archive {0} for job {1}".format(archive_index + 1, job_index + 1))
+      archfile = open(args.egs_dir + "/temp/" + args.prefix + "scp." + str(job_index + 1) + "." + str(archive_index + 1), "w")
+
+      this_egs = [] # this will be array of 2-tuples (lang-id start-frame num-frames)
+
+      num_egs = 0
+      while num_egs <= this_num_egs_per_archive:
+        rem_egs = sum(lang_len[i] for i in range(len(lang_len)))
+        if rem_egs > 0:
+          lang_id = RandomLang(lang_len, rem_egs, rand_select)
+          start_egs = lang2len[lang_id] - lang_len[lang_id]
+          this_egs.append((lang_id, start_egs, args.minibatch_size))
+          for scpline in range(args.minibatch_size):
+            print("{0} {1}".format(scp_files[lang_id].readline().splitlines()[0], lang_id), file = archfile)
+
+          lang_len[lang_id] = lang_len[lang_id] - args.minibatch_size
+          num_egs = num_egs + args.minibatch_size;
+          # If the num of remaining egs in each lang is less than minibatch_size,
+          # they are discarded.
+          if lang_len[lang_id] < args.minibatch_size:
+            lang_len[lang_id] = 0
+            print("Run out of data for language {0}".format(lang_id))
+        else:
+          print("Run out of data for all languages.")
+          break
+      all_egs.append(this_egs)
+      archfile.close()
+
+  # combine examples across all jobs correspond to each archive.
+  for archive in range(num_archives):
+    print("Processing archive {0} by combining all jobs.".format(archive + 1))
+    this_ranges = []
+    f = open(args.egs_dir + "/temp/" + args.prefix + "ranges." + str(archive + 1), "w")
```

Comment:
It might be better to name these as ranges.*.txt, output.*.txt etc.
end egs.*.scp